### PR TITLE
COMPRESS-382

### PR DESCRIPTION
**Pull request by tballison**
first draft of preventing OOM in LZMA

### CODE REVIEW

File: src/test/java/org/apache/commons/compress/compressors/DetectCompressorTestCase.java [hunk 1]
```diff
@@ -167,6 +171,26 @@ private String detect(String testFileName) throws IOException, CompressorExcepti
...
+        testMemoryLimit("COMPRESS-386");
+    }
+
+    private void testMemoryLimit(String fileName) throws IOException, CompressorException {
+        CompressorStreamFactory fac = new CompressorStreamFactory(true,
+                100);
+        try (InputStream is = new BufferedInputStream(
+                new FileInputStream(getFile(fileName)))) {
+            InputStream compressorInputStream = fac.createCompressorInputStream(is);
+            fail("Should have thrown CompressorMemoryLimitException");
```
Comment (kinow) [inline src/test/java/org/apache/commons/compress/compressors/DetectCompressorTestCase.java]: Maybe instead we could let the CompressorMemoryLimitException happen, and declare in the @Test annotation the expected exception.

Comment (bodewig) [inline src/test/java/org/apache/commons/compress/compressors/DetectCompressorTestCase.java]: +1 and make it two tests :-)

**Comment by tballison** [comment]
+1 to both suggestions.  Will do.  Thank you.

File: src/main/java/org/apache/commons/compress/compressors/lzma/LZMACompressorInputStream.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public LZMACompressorInputStream(final InputStream inputStream)
             in = new LZMAInputStream(inputStream, MemoryLimit.getMemoryLimitInKb());
         } catch (org.tukaani.xz.MemoryLimitException e) {
             //convert to commons-compress exception
-            throw new MemoryLimitException("exceeded calculated memory limit", e);
+            throw new MemoryLimitException(e.getMemoryNeeded(), e.getMemoryLimit());
```
Comment (kinow) [inline src/main/java/org/apache/commons/compress/compressors/lzma/LZMACompressorInputStream.java]: Not looking further into the source code of both MemoryLimitException classes, couldn't this hide the root cause of the exception? As the original `e` exception is ignored?

Comment (garydgregory) [inline src/main/java/org/apache/commons/compress/compressors/lzma/LZMACompressorInputStream.java]: I always like linking to the cause exception...

Comment (tballison) [inline src/main/java/org/apache/commons/compress/compressors/lzma/LZMACompressorInputStream.java]: Y.  Good point.  I'll update our MemoryLimitException initializers to both a 2-parameter (needed, limit) and 3-parameter (needed, limit, e)?

Comment (kinow) [inline src/main/java/org/apache/commons/compress/compressors/lzma/LZMACompressorInputStream.java]: Sounds good. There are a few other places in this pull request where the root exception is discarded too.

Comment (tballison) [inline src/main/java/org/apache/commons/compress/compressors/lzma/LZMACompressorInputStream.java]: Y, any time we're wrapping tukaani's.

**Pull request merged**